### ISSUE TITLE
WX version compatibility for Gooey, Pip, Conda, and OS platforms 

### ISSUE THREAD

**Issue by chriskiehl** [chriskiehl/Gooey#661]
Placeholder for working out how best to structure Gooey's WX dep such that installs across different package managers, platforms, and OSs can all work (mostly) seamlessly, or at least have a useful escape hatch. 

**Problems:**

* WX's [minor versions are not stable](https://github.com/wxWidgets/Phoenix/issues/1612), as such, Gooey is very conservative with its deps and locks WX to a known [stable version](https://github.com/chriskiehl/Gooey/issues/549) 
* The WX version Gooey uses isn't available in [package managers like Conda](https://github.com/chriskiehl/Gooey/commit/bc495ddc72b2b67fe735d90a654b0be8250b09d5). 
* There doesn't appear to be a single blessed WX version which can satisfy all installation requirements (for instance, the previous 4.0.7 caused [issues on Linux](https://github.com/chriskiehl/Gooey/issues/643)) 

**Planned Solution:** 

Gooey will loosen its grip on developing against a single known stable WX version, and work with a sliding window starting from `4.1.0` up to current `4.1.1` (and onward as relevant).

**Comment by casperdcl** [comment]
> - The WX version Gooey uses isn't available in package managers like Conda.

I assume this is in reference to https://github.com/conda-forge/gooey-feedstock/issues/6 which is no longer true. All WX versions are on conda atm.

Also, as of now Gooey *insists* on installing `wxpython==4.1.0` (hardcoded in `setup.py`) which prohibits using any different version even if compatible. I'd suggest reverting it back to wxpython>=4.0.7 or even earlier if known to work.

> Going back through the issue history a bit, it looks like I misspoke. People had trouble getting 4.0.7 installed on Linux, which the 4.1.0 bump resolved. So, there was a driver there beyond just getting things up to date.

I think there's a fundamental difference in thinking here. No package (Gooey, etc) need concern itself about whether its dependencies are easy to install. It simply should state the minimum version required, leaving it up to the end user to decide what to install.

`pip install gooey["wx>=4.1.0"]` sounds as crazy as `pip install "gooey-psutil-5.4.2[wx>=4.1,pillow>=4.3,colored>=1.3.93,pygtrie>=2.3.3]"`

It sounds like Gooey's deps should definitely be changed from `wxpython==4.1.0` to `wxpython>=4.0.7`.
If a particualr user is having difficulty installing `wxpython==4.0.7` then they can simply `pip install gooey "wxpython>=4.1"`

**Comment by casperdcl** [comment]
P.S. I'm on linux and have installed `wxpython==4.1.0` via `conda`

**Comment by chriskiehl** [comment]
The difference in thinking probably comes from me wearing the maintainer hat. Incentives shift based on perspective. Supporting a range of potentially non-backwards compatible WX versions is much harder to do on my end versus having a single canonical "good" version. FWIW, writing code to patch around version incompatibilities is what led to WxPython taking its current 'break everything' approach. Keeping things backwards compatible across dependency versions when that dependency keeps breaking its API introduces a lot of cruft into everything.   

 >It simply should state the minimum version required (and incompatible versions), leaving it up to the end user's package manager to decide what to install. 

This ignores the messy reality in favor of the ideal. I'd _love_ to just set a minimum and be done with it. However, _by design_, WX does not follow semvar nor guarantee backwards compatible releases. This fact is what led to using a locked down 'stable' version in the first place. Having Gooey stop working for _everyone_ because wx broke an API during a minor point release is an obnoxious way to spend a weekend. 

All that said, I can just bite the bullet and make sure Gooey works well with 4.0.7 - 4.1.1 and then slowly extend the upper supported versions as they come available.

**Comment by casperdcl** [comment]
I don't follow. My last suggestion (`wxpython>=4.0.7,<=4.1`) is about not artificially making old versions incompatible. As you say, "then slowly extend the upper supported versions as they come available."

When you say "make sure Gooey works well with 4.0.7 - 4.1.1" - I was under the impression that it does, and there's no effort required there... If you know the current state of gooey is incompatible with 4.0.7 then please drop it (rather than bother fixing it). I'm finding fault with dropping it without reason.

**Comment by chriskiehl** [comment]
>My last suggestion (wxpython>=4.0.7,<=4.1) is about not artificially making old versions incompatible. As you say, "then slowly extend the upper supported versions as they come available." 

OK. After rereading your comments several more times, I realize I've been misunderstanding what you're suggesting as well as talking about something slightly different. 

1. 4.0.7 is _not_ compatible with current Gooey versions. I injected a lot of confusion into the topic by suggesting that as a 'free' option
2. Given the above, I was coming at this from a perspective of supporting it _in spite_ of its incompatibility (+ issues from future WX versions). So, this was all from the vantage point that, given that the world is not ideal, and WX releases are what they are, and people have issues installing certain WX versions, what could Gooey do to make installing everywhere easy -- even if that meant carrying along some cruft in its code base to enable other WX versions to be used. Probably a dumb thing to worry about, I guess. 

> No package (Gooey, etc) need concern itself about whether its dependencies are easy to install. It simply should state the minimum version required (and incompatible versions)

FWIW, I somehow missed the "and incompatible versions" bit which I why I was talking about the ticking time bomb which would come from WX under a balnket `>=` setup. In short, I was mistaken. Whoops! 

So, TL;DR here, I'll update setup.py/requirements to use `wxpython>=4.1.0,<=4.1.1` and run with that model going forward.

**Comment by casperdcl** [comment]
Ah ok great, I hadn't realised either that 4.0.7 actually isn't compatible. Sure, no problem - I don't expect anyone to "fix" that :)

**Comment by tadam98** [comment]
Hi,

Getting an error with widgets example. simple and subparser run fine.

Thanks,
Mickey
```
GooeyExamples$ python ./examples/widget_demo.py
Writing Build Config to: ./examples/gooey_config.json
Gtk-Message: 22:35:43.923: Failed to load module "canberra-gtk-module"
Traceback (most recent call last):
  File "./examples/widget_demo.py", line 83, in <module>
    main()
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/python_bindings/gooey_decorator.py", line 87, in inner2
    return payload(*args, **kwargs)
  File "./examples/widget_demo.py", line 74, in main
    my_cool_parser.parse_args()
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/python_bindings/gooey_parser.py", line 114, in parse_args
    return self.parser.parse_args(args, namespace)
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/python_bindings/gooey_decorator.py", line 82, in run_gooey
    application.run(build_spec)
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/gui/application.py", line 21, in run
    app = build_app(build_spec)
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/gui/application.py", line 30, in build_app
    gapp = GooeyApplication(merge(build_spec, imagesPaths))
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/gui/containers/application.py", line 43, in __init__
    self.configs = self.buildConfigPanels(self)
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/gui/containers/application.py", line 221, in buildConfigPanels
    for widgets in self.buildSpec['widgets'].values()]
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/gui/containers/application.py", line 221, in <listcomp>
    for widgets in self.buildSpec['widgets'].values()]
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/gui/components/config.py", line 16, in __init__
    self.layoutComponent()
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/gui/components/config.py", line 92, in layoutComponent
    self.makeGroup(self, sizer, item, 0, wx.EXPAND | wx.ALL, 10)
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/gui/components/config.py", line 136, in makeGroup
    widget = self.reifyWidget(parent, item)
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/gui/components/config.py", line 187, in reifyWidget
    widgetClass = getattr(widgets, item['type'])
AttributeError: module 'gooey.gui.components.widgets' has no attribute 'TimeChooser'

```

**Comment by tadam98** [comment]
Hi again,

I forced conda install -c conda-forge wspython=4.1.0

```
GooeyExamples$ python ./examples/widget_demo.py
Writing Build Config to: ./examples/gooey_config.json
Gtk-Message: 22:45:23.054: Failed to load module "atk-bridge"
Gtk-Message: 22:45:23.056: Failed to load module "canberra-gtk-module"
Traceback (most recent call last):
  File "./examples/widget_demo.py", line 83, in <module>
    main()
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/python_bindings/gooey_decorator.py", line 87, in inner2
    return payload(*args, **kwargs)
  File "./examples/widget_demo.py", line 74, in main
    my_cool_parser.parse_args()
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/python_bindings/gooey_parser.py", line 114, in parse_args
    return self.parser.parse_args(args, namespace)
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/python_bindings/gooey_decorator.py", line 82, in run_gooey
    application.run(build_spec)
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/gui/application.py", line 21, in run
    app = build_app(build_spec)
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/gui/application.py", line 30, in build_app
    gapp = GooeyApplication(merge(build_spec, imagesPaths))
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/gui/containers/application.py", line 42, in __init__
    self.header = FrameHeader(self, buildSpec)
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/gui/components/header.py", line 33, in __init__
    self.layoutComponent()
  File "/home/mickey/miniconda3/envs/blurmvp2.2/lib/python3.7/site-packages/gooey/gui/components/header.py", line 77, in layoutComponent
    PAD_SIZE)
wx._core.wxAssertionError: C++ assertion "!(flags & wxALIGN_CENTRE_HORIZONTAL)" failed at /home/conda/feedstock_root/build_artifacts/wxpython_1603889841103/work/ext/wxWidgets/src/common/sizer.cpp(2106) in DoInsert(): Horizontal alignment flags are ignored in horizontal sizers
```

**Comment by methanm** [comment]
I don't know whether this will be helpful, but in debugging some issues I found that I had multiple versions of Python installed on my system:  3.6 (64-bit) and 3.8 (32-bit).  This was due to some configuration challenges we had faced some months back in our pipeline.  Revisiting the issues produced some problems with wxPython, but the difficulties were resolved by removing and reinstalling the base installations of Python on our system (a W10 box).

**Comment by casperdcl** [comment]
FYI the conda installation was fixed for `gooey>=1.0.8` https://github.com/conda-forge/gooey-feedstock/issues/6

**Comment by chriskiehl** [comment]
Dependency version has been changed to `>=4.x`
